lenskit-core/src/main/java/org/lenskit/util/XoRoShiRo128PlusRandomGenerator.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * LensKit, an open-source toolkit for recommender systems.
+ * Copyright 2014-2017 LensKit contributors (see CONTRIBUTORS.md)
+ * Copyright 2010-2014 Regents of the University of Minnesota
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining
+ * a copy of this software and associated documentation files (the
+ * "Software"), to deal in the Software without restriction, including
+ * without limitation the rights to use, copy, modify, merge, publish,
+ * distribute, sublicense, and/or sell copies of the Software, and to
+ * permit persons to whom the Software is furnished to do so, subject to
+ * the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be
+ * included in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+ * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+package org.lenskit.util;
+
+import org.apache.commons.math3.random.AbstractRandomGenerator;
+
+import javax.inject.Inject;
+
+/**
+ * XoRoShiRo128+ random number generator. This is implemented from the public-domain sources by David Blackman
+ * and Sebastiano Vigna at <http://xoroshiro.di.unimi.it/xoroshiro128plus.c> and the notes at
+ * <http://xoroshiro.di.unimi.it/>.
+ */
+public class XoRoShiRo128PlusRandomGenerator extends AbstractRandomGenerator {
```

Comment:
Convince me we need this.

Reply:
I have had some nagging concerns regarding the suitability of standard (i.e. bad) RNGs for the kinds of purposes we put them to in LensKit, especially with high-sample-count operations like BPR training. No concrete evidence of a problem, but Java Random is a basic `rand`, and far behind the state of the art; XoRoShiRo+ is also supposed to be a lot faster.

I wanted to rule out low-quality training pair sampling as a cause of my observed BPR woes. I knew a utility package by the author of `fastutil` had another RNG that was claimed to be much better; unfortunately, it is license-incompatible, but Vigna has a public domain C implementation of it, which I ported to Java here.

Swapping out the RNG made no difference in the problems I was seeing. Possibly silly thing to try in the first place. However, since I wrote the code, keeping it around lets us test RNG impact on LensKit runs and evaluation. It is disabled by default - need to rebind `RandomGenerator` to make use of it.